### PR TITLE
feat: server side filtering

### PR DESCRIPTION
**Pull request by yioneko**
New options:
- `vtsls.experimental.completion.enableServerSideFuzzyMatch`: Enable fuzzy match on server side and filter out items which has no match
- `vtsls.experimental.completion.entriesLimit`: Limit number of completion items returned from server

This will improve performance of clients with slow completion filtering.

### CODE REVIEW

**Comment by rchl** [comment]
Two cents from a Sublime Text's user perspective about this general tsserver issue...

What is slow in ST is not so much JSON parsing or filtering of completions but rather providing to ST API a large number of completion objects (`CompletionItem` in ST API). This has to be transferred across process boundaries in ST which appear to have the biggest impact. Once that's done, the filtering speed is no problem.

So the issue is most noticeable after typing the first character at word boundaries. Especially in global scope. This can in some cases return 50MB of completions.

And so, with a fuzzy matching on the server side, the benefit in that case will be probably pretty small since we only have one character to match again so most of the completions will still be present. It could even make things slower since fuzzy matching is not free.

Not sure where other editor's bottlenecks are so maybe it will help those more.

File: packages/service/configuration.schema.json [hunk 1]
```diff
@@ -1048,6 +1048,19 @@
...
+      "type": "boolean",
+      "description": "Execute fuzzy match of completion items on server side. Enable this will help filter out useless completion items from tsserver."
+    },
+    "vtsls.experimental.completion.entriesLimit": {
+      "default": null,
+      "type": [
+        "number",
+        "null"
+      ],
+      "description": "Maximum number of completion entries to return. Recommend to toggle `enableServerSideFuzzyMatch` either to preserve items with higher accuracy."
```
Comment (rchl) [inline packages/service/configuration.schema.json:1063]: Probably should be "either" => "also"

**Comment by yioneko** [comment]
> And so, with a fuzzy matching on the server side, the benefit in that case will be probably pretty small since we only have one character to match again so most of the completions will still be present. It could even make things slower since fuzzy matching is not free.

That is what `vtsls.experimental.completion.entriesLimit` deals with. In fact, the fuzzy matching is used to provide more accurate filtering on server side. The idea is taken from https://github.com/rust-lang/rust-analyzer/pull/13986.

**Comment by rchl** [comment]
Well, I wouldn't personally use `entriesLimit` because it's "destructive". Once the entries are limited, those will no longer be suggested, even after expanding the query, which can be confusing and annoying if user expected to find completion that was filtered out. Even together with fuzzy matching a useful completion can be filtered out.

Using fuzzy matching alone should be OK (in corner cases like when user presses backspace it could still have the same issue) but as I said, it might not always provide the expected benefit.

**Comment by rchl** [comment]
Also, when comparing with rust-analyzer, it's important to also specify whether they set `isIncomplete` on the completions result or not. If they do then it shouldn't be an issue to limit completions but tsserver/vtsls doesn't do it (in most cases).

**Comment by yioneko** [comment]
`isIncomplete` will be set when limited: https://github.com/yioneko/vtsls/blob/ba9f9a472a63ecfe5df09610d8da455237d85e56/packages/service/src/shims/languageFeatures.ts#L576

But `rust-analyzer` seems not do that.

**Comment by rchl** [comment]
OK, that's good. I haven't tried the changes, probably should. :)

File: README.md [hunk 2]
```diff
@@ -85,9 +81,20 @@ Almost the same as the original VSCode extension, with a few additional settings
...
+Please see [Known issues](https://github.com/yioneko/vtsls/issues/26) first.
+
+### Bad performance of completion
+
+`tsserver` could throw out plenty of completion entries, most of them are globally accessible variables, modules or namespaces. Some LSP clients have poor performance for fuzzy matching or filtering, and cause noticeable delay of completion.
+
+Instead of switching client, some server configuration options could also make partial optimizations:
 
-See [Known issues](https://github.com/yioneko/vtsls/issues/26)
+- `vtsls.experimental.completion.enableServerSideFuzzyMatch`: before returning all the completion candidates from `tsserver`, the server will do fuzzy matching and filter out entries with no match. This can significantly reduce the number of invalid entries.
```
Comment (yioneko) [inline README.md:94]: Yes, the option alone has very limited effect on filtering. A better description is needed.

**Comment by KaranAhlawat** [comment]
Continuing from https://github.com/joaotavora/eglot/discussions/1178, I'd love to test the changes out. I will build from source and report my experiences here.

**Pull request merged**